pkg/loader/loader.go

```diff
@@ -341,13 +343,164 @@ func LoadRootsWithConfig(cfg *packages.Config, roots ...string) ([]*Package, err
+	//             integration tests that if this function's roots parameter was
+	//             omitted, then the loader would act against the current
+	//             directory. while the command line flags would prevent this,
+	//             for purposes of backwards-compatibility, the following check
+	//             is performed here instead of in the deepcopy test.
+	//
+	//             this is a reminder to sort this out and determine what the
+	//             expected, desired behavior should be when roots is omitted
+	if len(roots) == 0 {
+		roots = []string{"."}
+	}
+
+	// ensure cfg.Dir is absolute. it just makes the rest of the code easier
+	// to debug when there are any issues
+	if cfg.Dir != "" && !filepath.IsAbs(cfg.Dir) {
+		ap, err := filepath.Abs(cfg.Dir)
+		if err != nil {
+			return nil, err
+		}
+		cfg.Dir = ap
 	}
 
-	for _, rawPkg := range rawPkgs {
-		l.Roots = append(l.Roots, l.packageFor(rawPkg))
+	// store the value of cfg.Dir so we can use it later if it is non-empty.
+	// we need to store it now as the value of cfg.Dir will be updated by
+	// a loop below
+	cfgDir := cfg.Dir
+
+	// addNestedGoModulesToRoots is given to filepath.WalkDir
+	// and adds the directory part of p to the list of roots
+	// IFF p is the path to a file named "go.mod"
+	addNestedGoModulesToRoots := func(
+		p string,
+		d os.DirEntry,
+		e error) error {
+
+		if e != nil {
+			return e
+		}
+		if !d.IsDir() && filepath.Base(p) == "go.mod" {
+			roots = append(roots, filepath.Join(filepath.Dir(p), "..."))
+		}
+		return nil
+	}
+
+	// uniqueRoots is used to keep track of the discovered packages to be nice
+	// and try and prevent packages from showing up twice when nested module
+	// support is enabled. there is not harm that comes from this per se, but
+	// it makes testing easier when a known number of modules can be asserted
+	uniqueRoots := sets.String{}
+
+	// the basic loader logic is as follows:
+	//
+	//    1. iterate over the list of provided roots
+	//
+	//    2. if a root uses the nested path syntax, ex. ..., then walk the
+	//       root's descendants to search for any any nested Go modules, and if
+	//       found, load them to the list of roots
+	//
+	//    3. iterate over the list of updated roots
+	//
+	//    4. update the loader config's Dir property to be the directory element
+	//       of the current root
+	//
+	//    5. execute the loader on the base element of the current root, which
+	//       will be either "./." or "./..."
+	//
+	// the following range operation is parts 1-2
+	for i, r := range roots {
+		// clean up the root
+		r := filepath.Clean(r)
+
+		// get the absolute path of the root
+		if !filepath.IsAbs(r) {
+
+			// if the initial value of cfg.Dir was non-empty then use it when
+			// building the absolute path to this root. otherwise use the
+			// filepath.Abs function to get the absolute path of the root based
+			// on the working directory
+			if cfgDir != "" {
```

Comment:
so cfgDir might not be $PWD?

Reply:
Yep. See the public `LoadRoots` function, it just constructs an empty `Config` -- https://github.com/kubernetes-sigs/controller-tools/blob/cb13ac551a0599044e50ed756735a8f438a24631/pkg/loader/loader.go#L316-L324. 

The `LoadRootsWithConfig` is event documented as _for testing_ -- https://github.com/kubernetes-sigs/controller-tools/blob/cb13ac551a0599044e50ed756735a8f438a24631/pkg/loader/loader.go#L326-L332

It was only set explicitly in the past from a test that used it with a temp directory. I switched to using it from `os.Chdir` as it better supports parallel testing. 

FWIW, it is eventually set to `os.Getwd` if still empty by the time it reaches https://cs.opensource.google/go/x/tools/+/refs/tags/v0.1.10:go/packages/packages.go;l=563